components/forms/CreateTxForm/MsgForm/MsgExecuteContractForm.tsx

```diff
@@ -0,0 +1,219 @@
+import Select from "../../../inputs/Select";
+import StackableContainer from "../../../layout/StackableContainer";
+
+const JsonEditor = dynamic(() => import("../../../inputs/JsonEditor"), { ssr: false });
+
+const customDenomOption = { label: "Custom (enter denom below)", value: "custom" } as const;
+
+const getDenomOptions = (assets: ChainInfo["assets"]) => {
+  if (!assets?.length) {
+    return [customDenomOption];
+  }
+
+  return [...assets.map((asset) => ({ label: asset.symbol, value: asset })), customDenomOption];
+};
+
+interface MsgExecuteContractFormProps {
+  readonly fromAddress: string;
+  readonly setMsgGetter: (msgGetter: MsgGetter) => void;
+  readonly deleteMsg: () => void;
+}
+
+const MsgExecuteContractForm = ({
+  fromAddress,
+  setMsgGetter,
+  deleteMsg,
+}: MsgExecuteContractFormProps) => {
+  const { chain } = useChains();
+
+  const denomOptions = getDenomOptions(chain.assets);
+
+  const [contractAddress, setContractAddress] = useState("");
+  const [msgContent, setMsgContent] = useState("{}");
+  const [selectedDenom, setSelectedDenom] = useState(denomOptions[0]);
+  const [customDenom, setCustomDenom] = useState("");
+  const [amount, setAmount] = useState("0");
+
+  const [contractAddressError, setContractAddressError] = useState("");
+  const [customDenomError, setCustomDenomError] = useState("");
+  const [amountError, setAmountError] = useState("");
+
+  useEffect(() => {
+    setContractAddressError("");
+    setCustomDenomError("");
+    setAmountError("");
+
+    const isMsgValid = (): boolean => {
+      const addressErrorMsg = checkAddress(contractAddress, chain.addressPrefix);
+      if (addressErrorMsg) {
+        setContractAddressError(`Invalid address for network ${chain.chainId}: ${addressErrorMsg}`);
+        return false;
+      }
+
+      if (selectedDenom.value === customDenomOption.value && !customDenom) {
+        setCustomDenomError("Custom denom must be set because of selection above");
+        return false;
+      }
+
+      if (amount && Number(amount) < 0) {
+        setAmountError("Amount must be empty or a positive number");
+        return false;
+      }
+
+      if (selectedDenom.value === customDenomOption.value && !Number.isInteger(Number(amount))) {
+        setAmountError("Amount cannot be decimal for custom denom");
+        return false;
+      }
+
+      return true;
+    };
+
+    const denom =
+      selectedDenom.value === customDenomOption.value ? customDenom : selectedDenom.value.symbol;
+
+    const microCoin = (() => {
+      try {
+        return macroCoinToMicroCoin({ denom, amount }, chain.assets);
+      } catch {
+        return { denom, amount: "0" };
+      }
+    })();
+
+    const msgContentUtf8Array = (() => {
+      try {
+        // The JsonEditor does not escape \n or remove whitespaces, so we need to parse + stringify
+        return toUtf8(JSON.stringify(JSON.parse(msgContent)));
+      } catch {
+        return Uint8Array.from([]);
```

Comment:
Ah, okay, I see. `undefined` and empty bytes are the same thing at protobuf level, so using that does not help much.

Can we prevent that invalid JSON can be submitted? It would be bad if a user creates an empty msg because they have some typo in their JSON.